File: java/hap20/proj/main/SessionCrypto.java
package main;

import java.io.IOException;

import crypto.ChachaDecoder;
import crypto.ChachaEncoder;
import bouncy.Pack;

public class SessionCrypto
{
	private static final int MAX_ENCRYPTED_LENGTH = 1024;
	int inboundBinaryMessageCount = 0;
	int outboundBinaryMessageCount = 0;
	private byte[] write_key;
	private byte[] read_key;

	public SessionCrypto(byte[] write_key, byte[] read_key)
	{
		this.write_key = write_key;
		this.read_key = read_key;
	}

	private byte[] encryptChunk(byte[] from, int len, byte[] to, int offset) throws Exception
	{
		byte[] additionalData = new byte[2];

		byte[] chunk;
		if (offset == 0 && len < MAX_ENCRYPTED_LENGTH)
		{
			chunk = from;
		}
		else
		{
			chunk = new byte[len];
			for (int i = 0; i < len; ++i)
			{
				chunk[i] = from[i + offset];
			}
		}

		additionalData[0] = (byte) (chunk.length & 0xff);
		additionalData[1] = (byte) (chunk.length >> 8);

		byte[] nonce = Pack.longToLittleEndian(outboundBinaryMessageCount++);
		byte[] ciphertext = new ChachaEncoder(write_key, nonce).encodeCiphertext(chunk, additionalData);

		// ///////////////
		byte[] merge = new byte[to.length + ciphertext.length + 2];

		int i;
		for (i = 0; i < to.length; ++i)
		{
			merge[i] = to[i];
		}

		merge[i] = additionalData[0];
		merge[i + 1] = additionalData[1];

		for (i = 0; i < ciphertext.length; ++i)
		{
			merge[i + 2 + to.length] = ciphertext[i];
		}

		return merge;
	}
	
	private byte[] decryptChunk(byte[] from, int len, byte[] to, int offset) 
	{
		byte[] mac = new byte[16];
		byte[] cipherchunk = new byte[len];
		System.arraycopy(from, offset, cipherchunk, 0, len);
		System.arraycopy(from, offset + len, mac, 0, 16);
		byte[] additionalData = new byte[2];
		additionalData[0] = (byte) (len & 0xff);
		additionalData[1] = (byte) (len >> 8);

		byte[] plaintextChunk;
		try
		{
			byte[] nonce = Pack.longToLittleEndian(inboundBinaryMessageCount++);
			plaintextChunk = new ChachaDecoder(read_key, nonce).decodeCiphertext(mac, additionalData, cipherchunk);
		}
		catch (IOException e)
		{
			throw new RuntimeException(e);
		}
		
		// ///////////////
		byte[] merge = new byte[to.length + plaintextChunk.length];

		int i;
		for (i = 0; i < to.length; ++i)
		{
			merge[i] = to[i];
		}

		for (i = 0; i < plaintextChunk.length; ++i)
		{
			merge[i + to.length] = plaintextChunk[i];
		}

		return merge;		
		
	}
	
	public byte[] decrypt(byte[] ciphertext)
	{
		
		int offset = 0;
		byte[] result = new byte[0];
		while (offset < ciphertext.length)
		{
	        int len = ((ciphertext[offset + 1] & 0xff) << 8) + (ciphertext[offset + 0] & 0xff);
			System.out.println ("session decrypt: processing chunk len = " + len);
			result = decryptChunk(ciphertext, len, result, offset + 2);
			offset += len + 2 + 16;
		}

		return result;
	}
	


	public byte[] encrypt(byte[] plaintext) throws Exception
	{
		int offset = 0;
		byte[] result = new byte[0];
		while (offset < plaintext.length)
		{
			int len = plaintext.length - offset;
			if (len > MAX_ENCRYPTED_LENGTH)
			{
				len = MAX_ENCRYPTED_LENGTH;
			}
			System.out.println ("session encrypt: processing chunk len = " + len);
			result = encryptChunk(plaintext, len, result, offset);
			offset += len;
		}

		return result;
	}

}

File: cpp/async2/session_crypto.cpp
#include "session_crypto.h"

SessionCrypto::SessionCrypto(int16_t _max, int8_t _mac_size) 
: max_encrypted_length (_max), mac_size (_mac_size) 
{
    read_buf.data = 0;
    read_buf.length = 0;
    write_buf.data = 0;
    write_buf.length = 0;
    inboundBinaryMessageCount = 0;
    outboundBinaryMessageCount = 0;
    this->write_key = 0;
    this->read_key = 0;

}

buf& SessionCrypto::encrypt(const buf& in)
{
    return encrypt(in.data, in.length);
}


buf& SessionCrypto::encrypt(const uint8_t* in, int len)
{
    free(write_buf.data);
    int offset = 0;
    write_buf.length = 0;
    write_buf.data = 0;
    while (offset < len)
    {
        int chunk_len = len - offset;
        if (chunk_len > max_encrypted_length)
        {
            chunk_len = max_encrypted_length;
        }
        printf ("session encrypt: processing chunk len = %d\n", chunk_len);
        encryptChunk(&in[offset], chunk_len, write_buf.data, write_buf.length);
        offset += chunk_len;
    }

    return write_buf;
}

void SessionCrypto::encryptChunk(const uint8_t* from, int from_len, uint8_t*& to, int& to_len)
{
    uint8_t additionalData[2];
    additionalData[0] = from_len & 0xff;
    additionalData[1] = (uint8_t)(from_len >> 8);

//    uint8_t* nonce = Pack.longToLittleEndian(outboundBinaryMessageCount++);
//    uint8_t* ciphertext = new ChachaEncoder(write_key, nonce).encodeCiphertext(chunk, additionalData);

    const uint8_t* ciphertext = from;
    int ciphertext_len = from_len + mac_size;  //TODO has to have real ciphertext len (from_len + 16)

    to = (uint8_t*) realloc (to, to_len + ciphertext_len + 2);
    to[to_len] = additionalData[0];
    to[to_len + 1] = additionalData[1];
    memcpy (&to[to_len + 2], ciphertext, from_len);  //TODO has to be ciphertext_len
    //fake auth array of 16 elements
    for (uint8_t i = 0; i < mac_size; ++i)
    {
        to[to_len + from_len + 2 + i] = 255;
    }
    to_len += ciphertext_len + 2;
}


//TODO validate lengths in case of an incomplete message
bool SessionCrypto::decrypt(const buf& from, buf& out)
{
    free (read_buf.data);
    read_buf.data = 0;
    read_buf.length = 0;
    int offset = 0;
    while (offset < from.length)
    {
        int chunk_len = ((from.data[offset + 1] & 0xff) << 8) + (from.data[offset + 0] & 0xff);
        if (offset + chunk_len + 2 + mac_size > from.length || chunk_len > max_encrypted_length) 
        {
            printf ("decrypt: chunk len too long\n");
            return false;
        }
        printf ("session decrypt: processing chunk len = %d\n", chunk_len);
        decryptChunk(&from.data[offset + 2], chunk_len, read_buf.data, read_buf.length);
        offset += chunk_len + 2 + mac_size;
    }

    out = read_buf;
    return true;

}

void SessionCrypto::decryptChunk (const uint8_t* from, int from_len, uint8_t*& to, int& to_len)
{
    //uint8_t mac[mac_size];  //TODO depends on input parameters of our real decrypt code.
    //memcpy (mac, &from[from_len], mac_size); //so may not need this
    for (int i = 0; i < mac_size; ++i) 
    {
        printf ("%d ", from[from_len]);
    }
    printf ("\n");

//    uint8_t additionalData[2];
//    additionalData[0] = from_len & 0xff;
//    additionalData[1] = from_len >> 8;

    //    byte[] nonce = Pack.longToLittleEndian(inboundBinaryMessageCount++);
//    plaintextChunk = new ChachaDecoder(read_key, nonce).decodeCiphertext(mac, additionalData, cipherchunk);

    const uint8_t* plaintextChunk = from;
    to = (uint8_t*) realloc (to, to_len + from_len);
    memcpy (&to[to_len], plaintextChunk, from_len); //TODO plainchunk len
    to_len += from_len; //TODO plainchunk len
}

void SessionCrypto::setKeys(uint8_t* write_key, uint8_t* read_key)
{
    this->write_key = write_key;
    this->read_key = read_key;
}
